### DISCUSSION TITLE
Need assistance in how to "resume" a song playing

### DISCUSSION THREAD

**Discussion by bernarma** [missionpinball/missionpinball#144] [Q&A] [answered]
I am trying to play a sound that resumes at a specific time in seconds when it was last active. I am currently storing the seconds that the song has been playing in the player_config via a timer but I am unsure how to use that value as a parameter to the "start_at" in the sound_player configuration.

```
sound_player:
  mode_song1_started:
    song1:
      action: play
      bus: music #temporary workaround
      start_at: (current_player.active_song_seconds)
```

The error I get in the log files when attempting to start MPF is which I am assuming is due to the parsing of the configuration file before the configuration for a player can even initialise its value to 0 (see the snippet from my config file):

```
mpf.exceptions.config_file_error.ConfigFileError: Config File Error in ConfigValidator: Failed to load config player mode_song1_started:sounds with settings {'song1{current_player.active_song_seconds != None}': {'action': 'play', 'bus': 'music', 'start_at': '(current_player.active_song_seconds)', 'block': False, 'delay': None, 'volume': None, 'ducking': {}, 'pan': None, 'loops': None, 'priority': None}} >> Config validation error: Entry sound_player:sound_player:start_at = "(current_player.active_song_seconds)" is not valid. Cannot convert value to secs. Error Code: CFE-ConfigValidator-11 (https://missionpinball.org/logs)
```

```
player_vars:
  active_song_seconds:
    initial_value: 0 # not being initialised in time
```

I am using the latest pre-release version of MPF and Godot.

**Comment by atummons** [comment]
First thing is player_vars default to type string. Add an argument below initial_value of value_type: intThe error can’t convert the value (string) to what it needs (int).Sent from my iPhoneOn Aug 19, 2024, at 8:15 PM, Mark Bernardinis ***@***.***> wrote:﻿
I am trying to play a sound that resumes at a specific time in seconds when it was last active. I am currently storing the seconds that the song has been playing in the player_config via a timer but I am unsure how to use that value as a parameter to the "start_at" in the sound_player configuration.
sound_player:
  mode_song1_started:
    song1:
      action: play
      bus: music #temporary workaround
      start_at: (current_player.active_song_seconds)

The error I get in the log files when attempting to start MPF is which I am assuming is due to the parsing of the configuration file before the configuration for a player can even initialise its value to 0 (see the snippet from my config file):
mpf.exceptions.config_file_error.ConfigFileError: Config File Error in ConfigValidator: Failed to load config player mode_song1_started:sounds with settings {'song1{current_player.active_song_seconds != None}': {'action': 'play', 'bus': 'music', 'start_at': '(current_player.active_song_seconds)', 'block': False, 'delay': None, 'volume': None, 'ducking': {}, 'pan': None, 'loops': None, 'priority': None}} >> Config validation error: Entry sound_player:sound_player:start_at = "(current_player.active_song_seconds)" is not valid. Cannot convert value to secs. Error Code: CFE-ConfigValidator-11 (https://missionpinball.org/logs)

player_vars:
  active_song_seconds:
    initial_value: 0 # not being initialised in time

I am using the latest pre-release version of MPF and Godot.

**Reply by bernarma** [reply]
Thanks - I'll give it a try tomorrow, but I did think it defaulted to int as per the documentation (not string).

![image](https://github.com/user-attachments/assets/415d5395-354c-4a7e-8fd5-187d8b79024d)

https://missionpinball.org/latest/config/player_vars/

**Comment by atummons** [comment]
Ahh, good catch. I was looking at wrong thing. What it’s yelling at is that you have an int, but it’s not in a time format. I’m not sure when I will have time to look into this one, but can try later this week if no one else gets to it.Sent from my iPhoneOn Aug 20, 2024, at 3:49 AM, Mark Bernardinis ***@***.***> wrote:﻿
Thanks - I'll give it a try tomorrow, but I did think it defaulted to int as per the documentation (not string).
image.png (view on web)
https://missionpinball.org/latest/config/player_vars/

**Reply by bernarma** [reply]
Thanks... I can confirm when it initializes all config files (I am assuming at start-up) before a game can even be started, the value it is indicates that it is receiving is 'None' (no quotes, just the value).

**Comment by bernarma** [comment]
OK, I have gone through the code briefly (not debugging interactively) but it appears as though the start_at value cannot handle values that are dynamically calculated at runtime using variables. I may be wrong but that is what it looks like. The question is, is this something that should be supported and is a bug or something that is not support and is therefore a feature???

**Comment by avanwinkle** [comment]
Ah hah, yes the documentation looks misleading to say the value should be an `int` when it should actually be a time value (int, float, or string). As for dynamic values, that's not currently available for `start_at:` but _could_ theoretically be easy by changing the config value from `secs` to `template_secs`. Let me poke around with that and see!

**Comment by avanwinkle** [comment, marked as answer]
Okay, it was a bit more complex than I was hoping because players like `sound_player` use nested dictionaries to pass multiple triggers from a single event. On the plus side, this change will apply to all config players and all template types, so it could be a big win for the future.

Please take a look!

https://github.com/missionpinball/mpf/pull/1824

**Reply by bernarma** [reply]
Geez... you are a machine. I will take a look. I just got the dev version of MPF up and running and was going to make an attempt at fixing it but I too saw the complexity as you described. Lucky I didn't start since you did it... let me merge this in to my local copy and see if it works satisfactorily. Stay tuned.

**Reply by bernarma** [reply]
OK, I ran it based off this snippet and it resolved to 0 which is a positive sign. I'll continue to test but I think you nailed it.

```
sound_player:
 mode_song1_started:
   song1:
     action: play
     bus: music #temporary workaround
     start_at: 0 if player.active_song_seconds == None else player.active_song_seconds
```

**Reply by avanwinkle** [reply]
That's great! You shouldn't need the conditional actually, accessing any undefined player variable will return an initial value of zero. Glad it's working for you!
